### PR TITLE
[master] Add `etcd-member-management` controller to K3s

### PR DESCRIPTION
**Pull request by Oats87**
#### Proposed Changes ####
This PR adds a new etcd member management controller to K3s, as well as moves the existing etcd controller to a "etcd metadata" controller that is responsible for updating the labels/annotations of corresponding etcd node objects.

#### Types of Changes ####
Functional

#### Verification ####
1. Create a multi-node etcd-backed K3s cluster.
2. Perform an `etcdctl member list` command using a downloaded `etcdctl` binary from the etcd project on any of the `etcd` nodes similar to:
```
etcdctl --cacert=/var/lib/rancher/k3s/server/tls/etcd/server-ca.crt --cert=/var/lib/rancher/k3s/server/tls/etcd/client.crt --key=/var/lib/rancher/k3s/server/tls/etcd/client.key member list -w table
```
and observe that there should be N etcd members listed, where N is the number of etcd nodes in the cluster.
3. Annotate one of your etcd nodes (preferably not the "bootstrap" node) with the annotation `etcd.k3s.cattle.io/remove=true`. List/describe the annotations of the node and watch the `etcd.k3s.io/removed-node-name` annotation be set to the etcd node name that corresponded to the list of nodes in the `member list` command from above.
4. Re-run the `etcdctl member list` and observe that the etcd member is NOT listed in the list anymore. Additionally, looking at the logs of `k3s server` of the node that was removed, observe that it appears quite angry and is constantly spitting logs related to etcd server stopped and so forth.

These 4 steps cover the primary additional functionality of the etcd member management controller. To validate the other bugfixes (2 of them) that were included by this PR, you can:

1. Restart the `k3s server` process i.e. `systemctl restart k3s` on the etcd node that was removed. Observe that it should re-join the etcd cluster (as observed by `etcdctl member list`) BUT it should get a new `etcd` member name and ID. When you `kubectl describe` the node, observe that the `remove` annotation should be reset to `false`.
2. Observe that `k3s` should not automatically be restarted by `systemd` or `openrc` once the node is removed from the etcd cluster. The node should basically just go into a bad state and need to be manually fixed via `systemctl restart`

#### Linked Issues ####
https://github.com/k3s-io/k3s/issues/3973

#### User-Facing Change ####
There are no immediate user facing changes. Most users should not be using this annotation functionality for managing etcd members.

```release-note

```

### CODE REVIEW

File: pkg/etcd/metadata_controller.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	return node, nil
+}
+
+func (m *metadataHandler) handleSelf(node *v1.Node) (*v1.Node, error) {
+	if node.Annotations[NodeNameAnnotation] == m.etcd.name &&
+		node.Annotations[NodeAddressAnnotation] == m.etcd.address &&
+		node.Labels[EtcdRoleLabel] == "true" &&
+		node.Labels[ControlPlaneLabel] == "true" ||
```
Comment (dweomer) [inline pkg/etcd/metadata_controller.go:51]: Given that the kubelet interprets any value for this label as applying this role, should we consider testing for the presence of the label and ignoring the value?

**Review by brandond — APPROVED** [review]

File: pkg/etcd/etcd.go [hunk 2]
```diff
@@ -54,6 +55,10 @@ const (
 	defaultKeepAliveTimeout = 10 * time.Second
 
 	maxBackupRetention = 5
+
+	MasterLabel       = "node-role.kubernetes.io/master"
+	ControlPlaneLabel = "node-role.kubernetes.io/control-plane"
+	EtcdRoleLabel     = "node-role.kubernetes.io/etcd"
```
Comment (brandond) [inline pkg/etcd/etcd.go:61]: We already have these as consts at https://github.com/k3s-io/k3s/blob/699ea165231cf6d665425c50ff5c75e1831f4afb/pkg/server/server.go#L42-L44

File: pkg/etcd/member_controller.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+	node = node.DeepCopy()
+
+	if removalRequested, ok := node.Annotations[removalAnnotation]; ok {
+		if removed, ok := node.Annotations[removedNodeNameAnnotation]; ok {
+			// check to see if removed is true. if it is, nothing to do.
+			if currentNodeName, ok := node.Annotations[NodeNameAnnotation]; ok {
+				if currentNodeName != removed {
+					// If the current node ID is not the same as the removed node ID, reset the tainted annotation and removed node ID
+					logrus.Infof("Resetting removed node flag as removed node ID ( did not match current node ID")
```
Comment (brandond) [inline pkg/etcd/member_controller.go:53]:
```suggestion
					// If the current cluster member name is not the same as the removed member name, reset the tainted annotation and removed member name
					logrus.Infof("Resetting removed node flag as removed node name did not match current node name")
```

File: pkg/etcd/member_controller.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+					delete(node.Annotations, removedNodeNameAnnotation)
+					node.Annotations[removalAnnotation] = "false"
+					return e.nodeController.Update(node)
+				}
+				// this is the case where the current node ID matches the removed node ID. We have already removed the
+				// node, so no need to perform any action. Fallthrough to the non-op below.
+			}
+			// This is the edge case where the removed annotation exists, but there is not a current node ID annotation.
+			// This should be a non-op, as we can't remove the node anyway.
```
Comment (brandond) [inline pkg/etcd/member_controller.go:62]:
```suggestion
				// this is the case where the current member name matches the removed member name. We have already removed the
				// member, so no need to perform any action. Fallthrough to the non-op below.
			}
			// This is the edge case where the removed annotation exists, but there is not a current member name annotation.
			// This should be a non-op, as we can't remove the member anyway.
```

**Review by dweomer — APPROVED** [review]

**Review by brandond — CHANGES_REQUESTED** [review]
A couple nits on use of node id vs member name.
etcd has members with name and id; the name and the id are not the same thing. We should be consistent about which one we're using. We should also be consistent about referring to Kubernetes nodes vs etcd members.

File: pkg/etcd/metadata_controller.go (hunk 1 above)
Comment (Oats87) [reply pkg/etcd/metadata_controller.go:51]: Potentially! Is there anywhere that the kubelet's behavior is defined for this? I know we just discussed this on the system-upgrade-controller project. Also, this is coming from copy-pasta (from my renaming of the controller)

File: pkg/etcd/member_controller.go (hunk 3 above)
Comment (Oats87) [reply pkg/etcd/member_controller.go:53]: Nice catch, thanks!

File: pkg/etcd/member_controller.go (hunk 4 above)
Comment (Oats87) [reply pkg/etcd/member_controller.go:62]: Nice catch, thanks!

**Review by brandond — COMMENTED** [review]
LGTM except for the duplicate consts; can we reuse those?

**Comment by Oats87** [comment]
> LGTM except for the duplicate consts; can we reuse those?

When using those constants, I run into an import cycle. Not sure if we want to move though

**Comment by brandond** [comment]
> When using those constants, I run into an import cycle. Not sure if we want to move those to a more "universal" point

Ugh yeah, it looks like we also have another copy in servicelb as well. I guess we can clean it up later.

**Review by brandond — APPROVED** [review]

**Review by galal-hussein — APPROVED** [review]
LGTM looks great thanks

**Pull request merged**